x/ccv/provider/keeper/proposal.go

```diff
@@ -21,17 +24,90 @@ import (
 // If the spawn time has already passed, then set the consumer chain. Otherwise store the client
 // as a pending client, and set once spawn time has passed.
 func (k Keeper) CreateConsumerChainProposal(ctx sdk.Context, p *types.CreateConsumerChainProposal) error {
-	if ctx.BlockTime().After(p.SpawnTime) {
-		return k.CreateConsumerClient(ctx, p.ChainId, p.InitialHeight)
+	if !ctx.BlockTime().Before(p.SpawnTime) {
+		return k.CreateConsumerClient(ctx, p.ChainId, p.InitialHeight, p.LockUnbondingOnTimeout)
+	}
+
+	k.SetPendingClientInfo(ctx, p)
+	return nil
+}
+
+// StopConsumerChainProposal stops a consumer chain and released the outstanding unbonding operations.
+// If the stop time hasn't already passed, it stores the proposal as a pending proposal.
+func (k Keeper) StopConsumerChainProposal(ctx sdk.Context, p *types.StopConsumerChainProposal) error {
+
+	if !ctx.BlockTime().Before(p.StopTime) {
+		return k.StopConsumerChain(ctx, p.ChainId, false, true)
+	}
+
+	k.SetPendingStopProposal(ctx, p.ChainId, p.StopTime)
+	return nil
+}
+
+// StopConsumerChain cleans up the states for the given consumer chain ID and, if the given lockUbd is false,
+// it completes the outstanding unbonding operations lock by the consumer chain.
+func (k Keeper) StopConsumerChain(ctx sdk.Context, chainID string, lockUbd, closeChan bool) (err error) {
+
+	// clean up states
+	k.DeleteConsumerClient(ctx, chainID)
+	k.DeleteLockUnbondingOnTimeout(ctx, chainID)
+
+	// close channel and delete the mappings between chain ID and channel ID
+	if channelID, found := k.GetChainToChannel(ctx, chainID); found {
+		if closeChan {
+			k.CloseChannel(ctx, channelID)
+		}
+		k.chanCloseInit(ctx, channelID)
```

Comment:
Shouldn't this line be inside the conditional as well?

Reply:
Definitely. Thanks!